x/fees/keeper/msg_server.go

```diff
@@ -0,0 +1,155 @@
+package keeper
+
+import (
+	"context"
+
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	sdkerrors "github.com/cosmos/cosmos-sdk/types/errors"
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/crypto"
+
+	"github.com/tharsis/evmos/v3/x/fees/types"
+)
+
+var _ types.MsgServer = &Keeper{}
+
+// RegisterDevFeeInfo registers a contract to receive transaction fees
+func (k Keeper) RegisterDevFeeInfo(
+	goCtx context.Context,
+	msg *types.MsgRegisterDevFeeInfo,
+) (*types.MsgRegisterDevFeeInfoResponse, error) {
+	ctx := sdk.UnwrapSDKContext(goCtx)
+	if !k.isEnabled(ctx) {
+		return nil, sdkerrors.Wrapf(types.ErrInternalFee, "fees module is not enabled")
+	}
+
+	contract := common.HexToAddress(msg.ContractAddress)
+	if k.IsFeeRegistered(ctx, contract) {
+		return nil, sdkerrors.Wrapf(sdkerrors.ErrInvalidRequest, "contract %s is already registered", contract)
+	}
+
+	var withdrawal sdk.AccAddress
+	deployer, _ := sdk.AccAddressFromBech32(msg.DeployerAddress)
+	derivedContractAddr := common.BytesToAddress(deployer)
+
+	if msg.WithdrawAddress != "" {
+		withdrawal, _ = sdk.AccAddressFromBech32(msg.WithdrawAddress)
+	}
+
+	// the contract can be directly deployed by an EOA or created through one
+	// or more factory contracts. If it was deployed by an EOA account, then
+	// msg.Nonces contains the EOA nonce for the deployment transaction.
+	// If it was deployed by one or more factories, msg.Nonces contains the EOA
+	// nonce for the origin factory contract, then the nonce of the factory
+	// for the creation of the next factory/contract.
+	for _, nonce := range msg.Nonces {
+		derivedContractAddr = crypto.CreateAddress(derivedContractAddr, nonce)
+	}
+
+	if contract != derivedContractAddr {
```

Comment:
this can succeed if `msg.ContractAddress = msg.Deployer`  and `len(msg.Nonces) = 0`. Are we checking that `len(msg.Nonces) > 0` in `msg.ValidateBasic`?

Reply:
The deployer address is checked by the cosmos sdk signature logic here: https://github.com/tharsis/evmos/pull/436/files#diff-cf04cadda8309151fbe358428c4b3c20b5baeda8f10fcff11ae4d546ecf39b57R68-R75. Isn't this enough to avoid the situation you described?

If it is enough, I can add a comment explaining the above. But do you still want me to add the following checks:
1. len(msg.Nonces) > 0
2. deployer address is not a contract

Reply:
Checks added in https://github.com/tharsis/evmos/pull/436/commits/404b219ca3740a5fb46595e330fc9817ceff2584